xapian-core/weight/bm25plusweight.cc

```diff
@@ -0,0 +1,157 @@
+ * @brief Xapian::BM25PlusWeight class - the BM25+ probabilistic formula
+ */
+/* Copyright (C) 2016  Vivek Pal
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include "xapian/weight.h"
+
+#include "debuglog.h"
+#include "omassert.h"
+#include "serialise-double.h"
+
+#include "xapian/error.h"
+
+#include <algorithm>
+#include <cmath>
+
+using namespace std;
+
+namespace Xapian {
+
+BM25PlusWeight *
+BM25PlusWeight::clone() const
+{
+    return new BM25PlusWeight(param_k1, param_k2, param_k3, param_b,
+			      param_min_normlen, param_delta);
+}
+
+void
+BM25PlusWeight::init(double factor)
+{
+    Xapian::doccount tf = get_termfreq();
+
+    double tw = 0;
+    // BM25+ formula gives termweight = (total_no_of_docs + 1) / tf
+    if (tf != 0) tw = (get_collection_size() + 1) / tf;
+
+    AssertRel(tw,>,0);
+
+    if (tw < 2) tw = tw * 0.5 + 1;
+    termweight = log(tw) * factor;
+    if (param_k3 != 0) {
+	double wqf_double = get_wqf();
+	termweight *= (param_k3 + 1) * wqf_double / (param_k3 + wqf_double);
+    }
+    termweight *= (param_k1 + 1);
+
+    LOGVALUE(WTCALC, termweight);
+
+}
+
+string
+BM25PlusWeight::name() const
+{
+    return "Xapian::BM25PlusWeight";
+}
+
+string
+BM25PlusWeight::serialise() const
+{
+    string result = serialise_double(param_k1);
+    result += serialise_double(param_k2);
+    result += serialise_double(param_k3);
+    result += serialise_double(param_b);
+    result += serialise_double(param_min_normlen);
+    result += serialise_double(param_delta);
+    return result;
+}
+
+BM25PlusWeight *
+BM25PlusWeight::unserialise(const string & s) const
+{
+    const char *ptr = s.data();
+    const char *end = ptr + s.size();
+    double k1 = unserialise_double(&ptr, end);
+    double k2 = unserialise_double(&ptr, end);
+    double k3 = unserialise_double(&ptr, end);
+    double b = unserialise_double(&ptr, end);
+    double min_normlen = unserialise_double(&ptr, end);
+    double delta = unserialise_double(&ptr, end);
+    if (rare(ptr != end))
+	throw Xapian::SerialisationError("Extra data in BM25PlusWeight::unserialise()");
+    return new BM25PlusWeight(k1, k2, k3, b, min_normlen, delta);
+}
+
+double
+BM25PlusWeight::get_sumpart(Xapian::termcount wdf, Xapian::termcount len,
+							Xapian::termcount) const
```

Comment:
There's too much indentation here - `Xapian::termcount` should be aligned with the `Xapian::termcount` at the start of the parameter list on the line above.